### PR TITLE
Change tags of parallel_for.

### PR DESCRIPTION
**Pull request by Rombur**

### CODE REVIEW

File: packages/Search/examples/bvh_driver/bvh_driver.cpp [hunk 1]
```diff
@@ -146,9 +146,9 @@ int main_( Teuchos::CommandLineProcessor &clp, int argc, char *argv[] )
         Kokkos::deep_copy( k, k_host );
 
         Kokkos::View<details::Nearest *, DeviceType> nearest_queries(
-            "neatest_queries", n_points );
+            "nearest_queries", n_points );
         Kokkos::parallel_for(
-            "register_nearest_queries",
+            "bvh_driver::main_() register_nearest_queries",
```
Comment (aprokop) [inline packages/Search/examples/bvh_driver/bvh_driver.cpp]: Should we use some version of `__func__`?

**Comment by aprokop** [comment]
I would argue that the names of Kokkos views and parallel regions should not have any spaces and parentheses in them, only [a-z], [A-Z], [0-9] and _. This would make later parsing easier.

File: packages/Search/examples/bvh_driver/bvh_driver.cpp (hunk 1 above)
Comment (sslattery) [inline packages/Search/examples/bvh_driver/bvh_driver.cpp]: I think I'm indifferent as long as we have a naming convention that makes it clear where the kernel is located

**Comment by Rombur** [comment]
ready for review @aprokop

**Comment by aprokop** [comment]
Neat! Thanks, @Rombur. Couple minor nitpicks:
1. Can we change it from `REGISTER_PARALLEL` (what do we register?) to something like `KOKKOS_NAME` or `REGION_NAME`? 
2. I assume that if we ever want to run on Windows (and if Kokkos ever does) we'll need to add an `#ifdef` clause
3. Have you checked whether CUDA produces meaningful names?

Don't know if we should, but could also write whitespace stripping/replacing on `__func__`. Really, would need to see what output looks like.

**Comment by Rombur** [comment]
I have changed the name of the macro. The names produced with CUDA are the same as without CUDA.

**Comment by aprokop** [comment]
This is kind of stuff I'm getting right now with CUDA:
```
                                                                      calculate_bouding_of_the_scene         0.07959                    1         0.07959  52.941   3.086
                                                      calculateBoundingBoxescalculate_bounding_boxes         0.02646                    1         0.02646  17$600   1.026
     N6Kokkos7BinSortINS_4ViewIPjJNS_6DeviceINS_4CudaENS_12CudaUVMSpaceEEEEEENS_7BinOp1DIS7_EES6_jEE         0.01954                    4         0.00488  12$997   0.758
                                                                               BVHset_bounding_boxes         0.00825                    1         0.00825   5$
487   0.320
                                                                 generateHierarchygenerate_hierarchy         0.00763                    1         0.00763   5$
075   0.296
                                                                assignMortonCodesassign_morton_codes         0.00447                    1         0.00447   2$
977   0.174
N6Kokkos7BinSortINS_4ViewIPjJNS_6DeviceINS_4CudaENS_12CudaUVMSpaceEEEEEENS_7BinOp1DIS7_EES6_jE20copy_permute_functorINS1_IPiJNS_10LayoutLeftES6_EEES7_NS1_ISC$
JS6_EEEEE         0.00075                    1         0.00075   0.497   0.029
N6Kokkos7BinSortINS_4ViewIPjJNS_6DeviceINS_4CudaENS_12CudaUVMSpaceEEEEEENS_7BinOp1DIS7_EES6_jE20copy_permute_functorINS1_IS2_JNS_10LayoutLeftES6_EEES7_S7_EE  
       0.00074                    1         0.00074   0.495   0.029
N6Kokkos7BinSortINS_4ViewIPjJNS_6DeviceINS_4CudaENS_12CudaUVMSpaceEEEEEENS_7BinOp1DIS7_EES6_jE12copy_functorIS7_NS1_IS2_JNS_10LayoutLeftES6_EEEEE         0.0$
051                    1         0.00051   0.341   0.020
N6Kokkos7BinSortINS_4ViewIPjJNS_6DeviceINS_4CudaENS_12CudaUVMSpaceEEEEEENS_7BinOp1DIS7_EES6_jE12copy_functorINS1_IPiJS6_EEENS1_ISC_JNS_10LayoutLeftES6_EEEEE  
       0.00051                    1         0.00051   0.338   0.020
            N6Kokkos4Impl15min_max_functorINS_4ViewIPjJNS_6DeviceINS_4CudaENS_12CudaUVMSpaceEEEEEEEE         0.00041                    1         0.00041   0$
275   0.016
                                                                                    querysecond_pass         0.00035                    1         0.00035   0$
232   0.013
                                                                                      BVHset_indices         0.00029                    1         0.00029   0$
194   0.011
                                                              calculateBoundingBoxesfill_ready_flags         0.00028                    1         0.00028   0$
186   0.011
```
So, first of all, we need to do something like `__func__ + ": " + std::string(x)`.

**Comment by aprokop** [comment]
So, first of all, we need to do something like `__func__ + ": " + std::string(x)`. Second, it seems that `__func__` does not actually include the class name so it is not sufficient.

**Comment by sslattery** [comment]
Should we have a macro for this then?

**Comment by aprokop** [comment]
With
```c++
#define REGION_NAME(x) __PRETTY_FUNCTION__+std::string(": ")+std::string(x)
```
the output has what we need
```
                                                                     calculate_bouding_of_the_scene         0.08116                    1         0.08116  53.422   3.185
static void DataTransferKit::Details::TreeConstruction<DeviceType>::calculateBoundingBoxes(Kokkos::View<DataTransferKit::Node*, DeviceType>, Kokkos::View<Data
TransferKit::Node*, DeviceType>) [with DeviceType = Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>]: calculate_bounding_boxes         0.02649                    1         0.02649  17.441   1.040
     N6Kokkos7BinSortINS_4ViewIPjJNS_6DeviceINS_4CudaENS_12CudaUVMSpaceEEEEEENS_7BinOp1DIS7_EES6_jEE         0.01949                    4         0.00487  12.
832   0.765
DataTransferKit::BVH<DeviceType>::BVH(Kokkos::View<const DataTransferKit::Box*, DeviceType>) [with DeviceType = Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>]: set_bounding_boxes         0.00824                    1         0.00824   5.423   0.323
static DataTransferKit::Node* DataTransferKit::Details::TreeConstruction<DeviceType>::generateHierarchy(Kokkos::View<unsigned int*, Device>, Kokkos::View<DataTransferKit::Node*, DeviceType>, Kokkos::View<DataTransferKit::Node*, DeviceType>) [with DeviceType = Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>]: gen
erate_hierarchy         0.00764                    1         0.00764   5.029   0.300
static void DataTransferKit::Details::TreeConstruction<DeviceType>::assignMortonCodes(Kokkos::View<const DataTransferKit::Box*, DeviceType>, Kokkos::View<unsigned int*, Device>, const DataTransferKit::Box&) [with DeviceType = Kokkos::Device<Kokkos::Cuda, Kokkos::CudaUVMSpace>]: assign_morton_codes         0.00446  
                  1         0.00446   2.937   0.175
N6Kokkos7BinSortINS_4ViewIPjJNS_6DeviceINS_4CudaENS_12CudaUVMSpaceEEEEEENS_7BinOp1DIS7_EES6_jE20copy_permute_functorINS1_IPiJNS_10LayoutLeftES6_EEES7_NS1_ISC_
JS6_EEEEE         0.00077                    1         0.00077   0.504   0.030
N6Kokkos7BinSortINS_4ViewIPjJNS_6DeviceINS_4CudaENS_12CudaUVMSpaceEEEEEENS_7BinOp1DIS7_EES6_jE20copy_permute_functorINS1_IS2_JNS_10LayoutLeftES6_EEES7_S7_EE  
       0.00075                    1         0.00075   0.491   0.029
N6Kokkos7BinSortINS_4ViewIPjJNS_6DeviceINS_4CudaENS_12CudaUVMSpaceEEEEEENS_7BinOp1DIS7_EES6_jE12copy_functorINS1_IPiJS6_EEENS1_ISC_JNS_10LayoutLeftES6_EEEEE  
       0.00051                    1         0.00051   0.335   0.020
N6Kokkos7BinSortINS_4ViewIPjJNS_6DeviceINS_4CudaENS_12CudaUVMSpaceEEEEEENS_7BinOp1DIS7_EES6_jE12copy_functorIS7_NS1_IS2_JNS_10LayoutLeftES6_EEEEE         0.00
050                    1         0.00050   0.331   0.020
            N6Kokkos4Impl15min_max_functorINS_4ViewIPjJNS_6DeviceINS_4CudaENS_12CudaUVMSpaceEEEEEEEE         0.00040                    1         0.00040   0.
265   0.016
void DataTransferKit::BVH<DeviceType>::query(Kokkos::View<Query*, DeviceType>, Kokkos::View<int*, DeviceType>&, Kokkos::View<int*, DeviceType>&) const 
...
```
But this is quite verbose, and I am no longer certain this is good.

**Comment by sslattery** [comment]
Cool - we should put that in utils or something

**Comment by aprokop** [comment]
Some discussions on [stackoverflow](https://stackoverflow.com/questions/42758832/usage-of-function-pretty-function-in-non-gcc-compiler) and [cplusplus](http://www.cplusplus.com/forum/general/179020/) about portability. It seems like people agree that the most portable way is to use `BOOST_CURRENT_FUNCTION`. 

Just tried out with the following patch:
```diff
diff --git a/packages/Utils/src/DTK_ConfigDefs.hpp b/packages/Utils/src/DTK_ConfigDefs.hpp
index 5bf7248..9318977 100644
--- a/packages/Utils/src/DTK_ConfigDefs.hpp
+++ b/packages/Utils/src/DTK_ConfigDefs.hpp
@@ -19,6 +19,7 @@
 
 #include <cstdint>
 #include <string>
+#include <boost/current_function.hpp>
 
 namespace DataTransferKit
 {
@@ -34,7 +35,7 @@ using LocalOrdinal = unsigned int;
 using GlobalOrdinal = uint64_t;
 
 // clang-format off
-#define REGION_NAME(x) __func__+std::string(x)
+#define REGION_NAME(x) BOOST_CURRENT_FUNCTION+std::string(":")+std::string(x)
 // clang-format on
 
 //---------------------------------------------------------------------------//
```
Works fine with both CUDA and OpenMP. Would be nice to try it out using intel compiler. @Rombur should we try it out on Titan?

**Comment by aprokop** [comment]
I'd like to have an informal poll here. We have few options here:
- Use @Rombur's original approach with hand-written names
- Use pretty function-like macros for wrapping names up

One is short, one is a lot longer. One is more automatic than the other. I'd like to see everybody's preferences. My original thinking was pretty function but the full output is really long.

**Comment by Rombur** [comment]
Let's just use boost, it makes the tag shorter to type. Not sure how you want to test that on Titan. Do you have an account?

**Comment by aprokop** [comment]
@Rombur I do, but never really used it. Actually, nevermind titan, I have access to some easier machines to use to try to compile DTK on.

**Comment by sslattery** [comment]
I would prefer to be more verbose if it is clean and gives us the information we need

**Comment by aprokop** [comment]
Was trying to compile using Intel 16.0.0.3, but got hit with this:
```
In file included from ../../DataTransferKit/packages/Search/src/DTK_LinearBVH_decl.hpp(20),
                 from DataTransferKit/packages/Search/src/DTK_LinearBVH_SERIAL.cpp(34):
../../DataTransferKit/packages/Search/src/details/DTK_DetailsPredicate.hpp(37): error: braces cannot be omitted for this subobject initializer
          : _query_point( {0., 0., 0.} )
                           ^
```

**Comment by Rombur** [comment]
@aprokop what's the gcc version you are using?

**Comment by aprokop** [comment]
gcc-4.7.2. This is the default at Sandia (though they are planning to update it to 4.9). 4.7.2 has been an officially supported version of Trilinos for a while.

**Comment by Rombur** [comment]
gcc-4.7.2 does not support all of C++11 you need gcc 4.8

**Comment by aprokop** [comment]
Unfortunately, 4.7.2 is what Trilinos has been stuck for a while. Maybe we should actually indicate the minimum version of compilers required for DTK. Could you or @dalg24 come up with a list? Or at least, indicate what is the minimum version of gcc that could be used?

**Comment by dalg24** [comment]
https://gcc.gnu.org/projects/cxx-status.html#cxx11

**Comment by aprokop** [comment]
OK, I don't think I have an easy access to Intel with newer gcc. I'm going to trust boost portability.

File: packages/Utils/src/DTK_ConfigDefs.hpp [hunk 2]
```diff
@@ -32,6 +34,10 @@ using LocalOrdinal = unsigned int;
 //! Global ordinal typedef.
 using GlobalOrdinal = uint64_t;
 
+// clang-format off
+#define REGION_NAME(x) BOOST_CURRENT_FUNCTION+std::string(":")+std::string(x)
```
Comment (dalg24) [inline packages/Utils/src/DTK_ConfigDefs.hpp]: Nice.

**Pull request merged**